### PR TITLE
Add gravity awful.placement + tests [WIP]

### PR DESCRIPTION
**Pull request by Elv13**
Begin work on #1607, the current patch isn't correct and it's only the beginning. It will need a *lot* more tests before I change all the placement code to support gravity.

The good news:

 * It can reproduce the problem
 * It isn't a "true" regression compared to 4.0, only a different bug (they both misbehave differently)

The bad news:

* If it ever "worked", it was because the code to remove the border was broken/deadcode
* I ran the tests with other gravities and saying "is isn't pretty" would be an euphemism...
* Fixing all the corner cases will take a lot of work
* It never worked, not even remotely
* `git grep -En '(border|bw)' | grep 2 | grep '*'` on Awesome 3.5 show many similar issues, so this is nothing new... but it's worst as I added many new placement function and none of them handle gravity correctly

### CODE REVIEW

File: lib/awful/placement.lua [hunk 1]
```diff
@@ -273,6 +273,19 @@ local outer_positions = {
...
+local gravity_map = {
+    north_west = {x = 0, y = 0, width = 2, height = 2},
+    north_east = {x = 2, y = 2, width = 0, height = 0}, --FIXME
+    south_east = {x = 2, y = 2, width = 0, height = 0}, --FIXME
+    south_west = {x = 2, y = 2, width = 0, height = 0}, --FIXME
+    center     = {x = 1, y = 1, width = 1, height = 1},
+    north      = {x = 1, y = 1, width = 1, height = 1}, --FIXME
+    south      = {x = 1, y = 1, width = 1, height = 1}, --FIXME
+    east       = {x = 1, y = 1, width = 1, height = 1}, --FIXME
+    west       = {x = 1, y = 1, width = 1, height = 1}, --FIXME
```
Comment (psychon) [inline lib/awful/placement.lua]: static gravity is missing from this list.

**Comment by psychon** [comment]
This PR tries to "work around" gravities in the Lua code. If we go down that route, we might end up with lots of complicated code, so I'll try to step back a bit:

There are two kinds of operations that need to be differentiated:
- Adding decorations to a client (gravity needs to be handled)
- Placing a client (gravity is not important)

"Adding decorations" is traditionally just titlebars, but for us that's also window borders since we use them as "some kind of" titlebar.

Current behaviour:
- Changing titlebar size and border size: The C code corrects for window gravity
- Placing the window (`c:geometry`): Gravity does not matter

As #1607 shows: Border size changes is not always related to gravity.

So, should gravity handling perhaps be removed from the C code (at least the API accessible to Lua) and into rc.lua instead?
Another idea: Move "border width is set to 0 for fullscreen clients" into the C code?
What other options do we have?

**Comment by Elv13** [comment]
@psychon The less changes to `awful.placement`, the better. If the C side could abstract all the corner cases, then that's good.

Do you want me to merge the 3 first commits, remove the other 2 and close this?

**Comment by psychon** [comment]
I tried to do something "like this" in: https://github.com/awesomeWM/awesome/pull/1614 (but incomplete since no Lua code was touched)
Dunno about merging this yet. That would add unused code to the test suite, wouldn't it?

**Comment by Elv13** [comment]
This is a dead end
